core/src/main/scala/api/sfs.scala

```diff
@@ -25,15 +30,19 @@ trait Filesystem {
    *  things too eagerly, nor caching data for too long.
    */
   def resolve(path: Path): Key
-  def metadata(key: Key): Metadata
-  def lookup(key: Key): Data
+  def metadata(key: Key): M[Metadata]
+  def lookup(key: Key): M[Data]
```

Comment:
What you say there about an asynchronous process to give you a better answer reminds me a lot about Stream.toString and cats.Show[Stream] with regards to referential transparency: https://github.com/typelevel/cats/pull/775